buildtest/buildsystem/base.py

```diff
@@ -0,0 +1,558 @@
+"""
+BuildConfig: loader and manager for build configurations, and schema validation
+Copyright (C) 2020 Vanessa Sochat.
+
+A build configuration file can define any of the following sections, which
+are added to the test script in the following order:
+
+ - pre_build
+ - build
+ - post_build
+ - pre_run
+ - run
+ - post_run
+
+A cd to the $TESTDIR is always the first step taken, and any defined
+sections after that are added if not empty.
+The following sections can be defined, and serve different purposes:
+
+ - env
+
+"""
+
+from copy import deepcopy
+import datetime
+import json
+import os
+import re
+import stat
+import subprocess
+import sys
+import yaml
+
+from jsonschema import validate
+from jsonschema.exceptions import ValidationError
+
+from buildtest.config import config_opts
+from buildtest.log import init_logfile, init_log
+from buildtest.defaults import (
+    BUILDTEST_SHELL,
+    TESTCONFIG_ROOT,
+    BUILDTEST_BUILD_LOGFILE,
+    build_sections,
+    supported_schemas,
+    variable_sections,
+)
+from buildtest.utils.file import create_dir
+from buildtest.utils.command import BuildTestCommand
+from buildtest.buildsystem.schemas.utils import (
+    load_schema,
+    load_recipe,
+    get_schemas_available,
+    here,
+)
+
+known_sections = variable_sections + build_sections
+
+
+class BuildConfig:
+    """A BuildConfig is a base class for a build test configuration.
+       The type (e.g., script) and version are derived from reading in
+       the file, and then matching to a buildtest schema, each of which is
+       developed at https://github.com/HPC-buildtest/schemas and added to
+       subfolders named accordingly under buildtest/tools/buildsystem/schemas.
+       The schema object can load in a general test configuration file
+       to validate it, and then match it to a schema available.
+       If the version of a schema is not specified, we use the latest.
+       If the schema fails validation, we also don't continue.
```

Comment:
bit wordy **we also don't continue**. We could just say: **If the schema fails validation check, then we stop immediately.**